ADR/pipeline-service/adr-0001.md

```diff
@@ -0,0 +1,46 @@
+# Service Account configuration for image builds in AppStudio 
+
+Date: Oct 5, 2022
+
+## Status
+
+Proposed
+
+## Context
+
+The `serviceAccount` used for running a `PipelineRun` workload in a workload cluster needs to have the right SCCs and linked credentials to be able to successfully execute a container image build. The linked credentials will mostly originate from the KCP workspace where the `PipelineRun` was synced from.
+Additionally, the SCC association needed for elevating capabilities required for image builds isn't something we'd prefer to expose the user to.
+
+
+
+## Decision
+
+* As a Pipeline Service platform feature, any service/user intending to build a container image would need to create/use the "image-builder" service account in the KCP workspace. 
+Use of any other service accounts will not lead to a successful assembling of the image.
+
+* The AppStudio Build controller would tap into this platform contract to create relevant PipelineRuns for building container images.
+
+## Consequences
+
+
+### Implementation - AppStudio Build Service Controller
+
+* The AppStudio Build controller creates an "image-builder" service account in the KCP workspace/namespace.
+* The AppStudio Build controller links the image pull secret ( previously created by SPI ) with the "image-builder" `serviceAccount` in the KCP workspace/namespace. Upon doing so, the 'updated' "image-builder" `serviceAccount` would get synced into the workload cluster.
+* The AppStudio Build controller" creates the PipelineRun referencing the "image-builder" service account as the one that the PipelineRun/TaskRun `Pods` would run as. Upong doing so, the PipelineRun would get synced into the Pipeline Service Workload cluster.
+* And that's it! Since the `[cluster]rolebinding` in the Workload cluster for all "image-builder" `serviceAccounts` would be pre-configured ( by the Pipeline Service ) with the SCC needed to build container images, the image should be successfully assembled. 
```

Comment:
I'm fine with implementing this solution as long as we have an exit strategy.

We're starting to care about the workloads we are running, and I wish that was not the case. What happens when we need to build/push something else that also requires this type of credentials? How many *-builder SA will we have to maintain in the future. Do we have any guarantee that it won't be more than one?

Reply:
@adambkaplan 's second bullet here, at least as I like to read into it, has already come up with jvm-build-service on KCP progress I have been able to make today.  Though the added permissions I needed were not related to pod security, but additional API access.

Specifically, I have add to add RBAC permission on the workload cluster to allow the 
- "build SA" to be able to update taskruns/status as part of our `maven-run : analyse-dependencies` step (you get this step when you use the HACBS bundle to build java)
- to allow the maven proxy / artifact cache, which ideally we like to run under the "build SA" given the common permission requirements, to see reasonably simple things like get/watch/list configmaps to obtain needed metadata

It occurred to me that these short term hacks I had to perform conceivably all under the umbrella of this design and what it is trying to achieve.  Or at least close enough we could stretch this design ever so slightly.

WDTY @sbose78 @adambkaplan 

@stuartwdouglas FYI

Reply:
Conceptually the RBAC for SCC and the RBAC to update taskrun status is the same - it is an elevated permission that applies to the workload cluster, but not KCP.

Reply:
Yep @adambkaplan 's  wording lines up with what I was trying to say previously.  And since it is conceptually the same, it would ideal if we can solve both concerns with this effort.

Reply:
@Roming22  I've added the note on exit strategy in the "Security considerations" section. I'd say, the Pipeline Service team takes the call on who can be allowed to use elevated privileges. We'll need to build the machinery for that. Consumers may choose to opt-in but in the end, whoever owns the workload cluster is the ultimate approver. As of today, it's going to be us. However, when customers bring their own workload clusters, we should provide them with the machinery to manage such privileges.

Reply:
^ We'd have to take a call on"when" we'd like to prioritize building the fancy machinery :)

Reply:
@gabemontero @adambkaplan Agreed.

I can almost see a related ADR cc @Roming22 :

"As a consumer Foo, I need special permissions on the service account that I'm going to use".

* What are the boundaries ? The service account permissions specific to the consumer cannot break out of the namespace in the workload cluster. In this case, the consumer wants to read/write TaskRuns in the same namespace which feels allow-able.
* How does the consumer define such a configuration ? Pipeline Service infra should provide a framework for specifying such custom ClusterRoles whose policies are local to a namespace.

Clearly, we shouldn't be proliferating our ClusterRole with unrelated rules, rather providing a framework to create new ones which would have to be "approved" into the system and then associated with new/existing service accounts.

Reply:
So @sbose78 @adambkaplan @Roming22 @jduimovich @stuartwdouglas I think the discussion in https://coreos.slack.com/archives/C032EJ007C0/p1666186257512109, which arose from our discussion in the cabal 2 days ago, where that cabal agenda item from @sbose78 cited this discussion thread, landed us in a tenable spot wrt jvm-build-service needs, though I feel we need to close the loop and reflect that outcome in your ADR here @sbose78 , and ultimately get the appropriate Jira's for @Roming22 's team (unless they already exist @Roming22 ?).

The high level outcome from https://coreos.slack.com/archives/C032EJ007C0/p1666186257512109:
- https://github.com/openshift-pipelines/pipeline-service-workspace-controller will sync an SA and linked secret containing the kubeconfig and token needed to communicate with the KCP api server
- jvm-build-service will use said SA/secret with the pipelineruns it generates for processing artifacts/dependencies
- those pipelineruns can then update jvmbuildservice defined custom resources with the deploy artifact information vs. the current updating of taskrun/status with results (where tekton then aggregates to pipelinerun/status, and the syncer currently sinks back to kcp)
- any RBAC for updating those customresources can now be defined on the kcp side vs. compute side, and hence make sense to be entirely included in the manifest list jvm-build-service would provide to  @rajivnathan 's https://issues.redhat.com/browse/ASC-181 as part of establishing a hacbs enabled workspace for a user.

And this alleviates the concern around jvm-build-service related manipulation of results off the pipelinerun which enterprise contract subsequently inspects, as that manipulation would cease.

My remaining questions
- any remaining concerns?
- is there just one SA with linked kcp token/kubeconfig secret and with workload cluster SCC related changes for image builds ?
- or do we want on SA for building images, and another SA for "kcp aware" pipelineruns like what jvm-build-service needs?

Reply:
cross referencing with https://coreos.slack.com/archives/C02CTEB3MMF/p1666786789821739

learned from @scoheb today that other controllers under the HACBS umbrella besides jvm build service also have the requirement of a KCP aware SA+secret so that pipelineruns created by KCP aware controllers can manipulate KCP only CRs.